File: programmers/정수 삼각형.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cstring>

using namespace std;
int cache[500][500];
vector<vector<int>> map;
int cal(int x,int y)
{
    if(y >= map.size()) return 0;
    int& result = cache[y][x];
    if(result != -1) return result;
    result = map[y][x];
    result += max(cal(x,y+1),cal(x+1,y+1));
    return result;
}

int solution(vector<vector<int>> triangle) {
    int answer = 0;
    map = triangle;
    memset(cache,-1,sizeof(cache));
    answer = cal(0,0);
    return answer;
}

File: JAVA/Programmers/정수삼각형.java
class Solution {
    static int dp[][];
    static int size;
    static int dfs(int y , int x,int[][]triangle){
        int result =  0;
        if(y == size) return 0;
        if(dp[y][x] != 0) return dp[y][x];
        result = Math.max(dfs(y+1,x,triangle),dfs(y+1,x+1,triangle));
        result += triangle[y][x];
        dp[y][x] = result;
        return result;
    }
    public int solution(int[][] triangle) {
        int answer = 0;
        size = triangle.length;
        dp = new int [size][size];
        answer = dfs(0,0,triangle);
        return answer;
    }
}
